envoy/config/filter/http/jwt_authn/v2/config.proto

```diff
@@ -0,0 +1,236 @@
+  // fetch JWKS.
+  oneof jwks_source_specifier {
+    option (validate.required) = true;
+
+    // JWKS can be fetched from remote server via HTTP/HTTPS.
+    // This field specifies the remote HTTP URI.
+    //
+    // Example:
+    //
+    // .. code-block:: yaml
+    //
+    //    remote_jwks:
+    //    - http_uri:
+    //      - uri: https://www.googleapis.com/oauth2/v1/certs
+    //        cluster: jwt.www.googleapis.com|443
+    //      cache_duration:
+    //      - seconds: 300
+    //
+    RemoteJwks remote_jwks = 3;
+
+    // JWKS is in local data source. It could be either:
+    // in a local file or embedded in the inline_string.
+    //
+    // Example: local file
+    //
+    // .. code-block:: yaml
+    //
+    //    local_jwks:
+    //    - filename: /etc/envoy/jwks/jwks1.txt
+    //
+    // Example: inline_string
+    //
+    // .. code-block:: yaml
+    //
+    //    local_jwks:
+    //    - inline_string: "ACADADADADA"
+    //
+    envoy.api.v2.core.DataSource local_jwks = 4;
+  }
+
+  // If false, the JWT is removed in the request after a success verification.
+  // If true, the JWT is not removed in the request. Default value is false.
+  bool forward_jwt = 5;
+
+  // If true, the request is allowed if this JWT verification fails.
+  // If false, a failed verification request will be rejected.
+  // Default value is false.
+  bool allow_failed_jwt = 6;
+
+  // Two fields below define where to extract the JWT from an HTTP request.
+  //
+  // If no explicit location is specified, the following default locations are
+  // tried in order:
+  //
+  // 1. The Authorization header using the Bearer schema.
+  //    See `here <https://tools.ietf.org/html/rfc6750#section-2.1>`_
+  //    Example::
+  //
+  //       Authorization: Bearer <token>.
+  //
+  // 2. `access_token` query parameter.
+  //    See `this <https://tools.ietf.org/html/rfc6750#section-2.3>`_
+  //
+
+  // Specify the HTTP headers to extract JWT token.
+  //
+  // For examples, following config::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_headers:
+  //   - name: x-goog-iap-jwt-assertion
+  //
+  // can be used to extract token from header::
+  //
+  //   x-goog-iap-jwt-assertion: <JWT>.
+  //
+  repeated JwtHeader jwt_headers = 7;
+
+  // JWT is sent in a query parameter. `jwt_params` represents the
+  // query parameter names.
+  //
+  // For example, if config is::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_params:
+  //   - jwt_token
+  //
+  // The JWT format in query parameter is::
+  //
+  //    /path?jwt_token=<JWT>
+  //
+  repeated string jwt_params = 8;
+
+  // This field specifies the header name to forward a successfully verified
+  // JWT payload to the backend. The forwarded data is::
+  //
+  //    base64_encoded(jwt_payload_in_JSON)
+  //
+  // If it is not specified, the payload will not be forwarded.
+  // If multiple JWT payloads needed to be forwarded, distinct header names
```

Comment:
This is still a bit unclear to me. Below is a singleton with only one header name, where are the distinct header names located in the config?

Reply:
distinct header names are for across different issuers.  The singlton is only for one issuer. A request may have multiple tokens from different issuers, they need different header name to carry their payload.

Reply:
I'm still a bit confused here. Can you provide an example?

Reply:
A request has two JWTs
header1: <JWT1>
header2: <JWT2>

Need to pass both JWT payload to backend.  In this case, "jwt_payload_header" field for JWT1 should be different with the one in JWT2.  Otherwise both will use the same header to pass JWT payload,  they will collide.